### PR TITLE
Make trivial thread-idx for degenerate statements without thread-idx.

### PR DESCRIPTION
**Pull request by zheng-xq**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #36563 Add sync-point insertions and block/thread local memory allocations
* **#36480 Make trivial thread-idx for degenerate statements without thread-idx.**

Differential Revision: [D20992505](https://our.internmc.facebook.com/intern/diff/D20992505)

### CODE REVIEW

File: torch/csrc/jit/tensorexpr/cuda_codegen.cpp [hunk 1]
```diff
@@ -469,6 +475,132 @@ std::string CudaCodeGen::GetUniqueFuncName(const std::string& func_prefix) {
   return func_prefix + "_" + std::to_string(value);
 }
 
+// Find all the statements that are not covered by any thread-idx axes,
+// and wrapp them under a trivial thread idx.
+class NoThreadIdxRewriter : public IRMutator {
+ private:
+  Stmt* rewrite(const std::vector<Stmt*>& stmts) {
+    std::vector<Stmt*> cloned_stmts(stmts.size());
+    for (int index = 0; index < stmts.size(); index++) {
```
Comment (ZolotukhinM) [inline torch/csrc/jit/tensorexpr/cuda_codegen.cpp:484]: Please fix the clang-tidy complaint :)

File: torch/csrc/jit/tensorexpr/cuda_codegen.cpp [hunk 2]
```diff
@@ -469,6 +475,132 @@ std::string CudaCodeGen::GetUniqueFuncName(const std::string& func_prefix) {
   return func_prefix + "_" + std::to_string(value);
 }
 
+// Find all the statements that are not covered by any thread-idx axes,
+// and wrapp them under a trivial thread idx.
```
Comment (ZolotukhinM) [inline torch/csrc/jit/tensorexpr/cuda_codegen.cpp:479]: Typo: wrapp.

File: torch/csrc/jit/tensorexpr/cuda_codegen.cpp [hunk 3]
```diff
@@ -469,6 +475,132 @@ std::string CudaCodeGen::GetUniqueFuncName(const std::string& func_prefix) {
...
+  Stmt* rewrite(const std::vector<Stmt*>& stmts) {
+    std::vector<Stmt*> cloned_stmts(stmts.size());
+    for (int index = 0; index < stmts.size(); index++) {
+      cloned_stmts[index] = Stmt::clone(stmts[index]);
+    }
+    Stmt* new_block = Block::make(cloned_stmts);
+    // Wrap the new block under a trivial thread-idx
+    //   for t in 0..1: // threadIdx
+    //     if (t < 1):
+    //       new_block
```
Comment (ZolotukhinM) [inline torch/csrc/jit/tensorexpr/cuda_codegen.cpp:491]: Not a concern for this PR, but this structure seems like an overkill for just setting a thread-idx for new_block. Is there another way we can represent it?

File: torch/csrc/jit/tensorexpr/cuda_codegen.cpp [hunk 4]
```diff
@@ -469,6 +475,132 @@ std::string CudaCodeGen::GetUniqueFuncName(const std::string& func_prefix) {
...
+    }
+
+    return new_for;
+  }
+
+  Stmt* mutate(const Block* v) override {
+    std::list<Stmt*> old_stmts = v->stmts();
+    std::vector<bool> need_rewrites(old_stmts.size());
+    std::vector<Stmt*> new_stmts(old_stmts.size());
+    ;
```
Comment (ZolotukhinM) [inline torch/csrc/jit/tensorexpr/cuda_codegen.cpp:528]: Nit: unnecessary ";"

File: torch/csrc/jit/tensorexpr/cuda_codegen.cpp [hunk 5]
```diff
@@ -469,6 +475,132 @@ std::string CudaCodeGen::GetUniqueFuncName(const std::string& func_prefix) {
...
+    // If nothing needs fix, return as it is
+    if (!any_need_fix) {
+      return (Stmt*)v;
+    }
+
+    // If all needs fix, check if parent is a GPU block
+    if (all_need_fix) {
+      Stmt* parent = v->get_parent();
+      For* loop_parent = dynamic_cast<For*>(parent);
+      if (loop_parent) {
```
Comment (ZolotukhinM) [inline torch/csrc/jit/tensorexpr/cuda_codegen.cpp:558]: Nit: can do "`if (loop_parent && loop_parent->loop_options().is_gpu_block_index()) {`

File: torch/csrc/jit/tensorexpr/cuda_codegen.cpp [hunk 6]
```diff
@@ -469,6 +475,132 @@ std::string CudaCodeGen::GetUniqueFuncName(const std::string& func_prefix) {
...
+        rewrite_stmts.push_back(Stmt::clone(new_stmts[start]));
+        start++;
+      }
+      int stop = start + 1;
+      while (stop < count && need_rewrites[stop]) {
+        stop++;
+      }
+
+      // Rewrite the stmts from [start, stop)
+      std::vector<Stmt*> stmts_block(
```
Comment (ZolotukhinM) [inline torch/csrc/jit/tensorexpr/cuda_codegen.cpp:583]: Nit: the name `stmts_block` is slightly misleading as it's not a `Block` (even though it's very similar by its meaning).

File: torch/csrc/jit/tensorexpr/cuda_codegen.cpp [hunk 7]
```diff
@@ -469,6 +475,132 @@ std::string CudaCodeGen::GetUniqueFuncName(const std::string& func_prefix) {
...
+          new_stmts.begin() + start, new_stmts.begin() + stop);
+      Stmt* rewritten_stmt = rewrite(stmts_block);
+      rewrite_stmts.push_back(rewritten_stmt);
+
+      start = stop;
+    }
+    Stmt* rewritten_block = Block::make(rewrite_stmts);
```
Comment (ZolotukhinM) [inline torch/csrc/jit/tensorexpr/cuda_codegen.cpp:590]: Maybe add an assert that rewrite_stmts.size() == new_stmts.size() == old_stmts.size()?

File: torch/csrc/jit/tensorexpr/cuda_codegen.cpp [hunk 8]
```diff
@@ -469,6 +475,132 @@ std::string CudaCodeGen::GetUniqueFuncName(const std::string& func_prefix) {
...
+    }
+
+    need_rewrite_ = false;
+    // If nothing needs fix, return as it is
+    if (!any_need_fix) {
+      return (Stmt*)v;
+    }
+
+    // If all needs fix, check if parent is a GPU block
+    if (all_need_fix) {
```
Comment (ZolotukhinM) [inline torch/csrc/jit/tensorexpr/cuda_codegen.cpp:555]: Why do we need to special-case this? Won't the logic below handle it as well?

**Review by ZolotukhinM — APPROVED** [review]
Looks good overall, some comments are inline!

File: test/cpp/tensorexpr/test_cuda.cpp [hunk 9]
```diff
@@ -477,6 +477,103 @@ void testCudaOneBlockMultiThreadGlobalReduce1() {
...
+  a_ref(0) = 0;
+  for (int i = 0; i < 2; i++) {
+    a_ref(0) += i;
+  }
+  a_ref(1) = a_ref(0) + 1;
+  for (int i = 0; i < N; i++) {
+    b_ref(i) = i;
+  }
+
+  float* a_dev = nullptr;
```
Comment (nickgg) [inline test/cpp/tensorexpr/test_cuda.cpp:555]: Running the codegen is not sufficient here, I think. Can you verify the correct code was generated in a literal way?

File: torch/csrc/jit/tensorexpr/cuda_codegen.cpp (hunk 1 above)
Comment (zheng-xq) [reply torch/csrc/jit/tensorexpr/cuda_codegen.cpp]: Done

File: torch/csrc/jit/tensorexpr/cuda_codegen.cpp (hunk 2 above)
Comment (zheng-xq) [reply torch/csrc/jit/tensorexpr/cuda_codegen.cpp]: Done

File: torch/csrc/jit/tensorexpr/cuda_codegen.cpp (hunk 3 above)
Comment (zheng-xq) [reply torch/csrc/jit/tensorexpr/cuda_codegen.cpp:491]: Add a note that this serves too purposes: 1. setting the mask; 2. setting the sync-point boundary.

File: torch/csrc/jit/tensorexpr/cuda_codegen.cpp (hunk 4 above)
Comment (zheng-xq) [reply torch/csrc/jit/tensorexpr/cuda_codegen.cpp]: done

File: torch/csrc/jit/tensorexpr/cuda_codegen.cpp (hunk 8 above)
Comment (zheng-xq) [reply torch/csrc/jit/tensorexpr/cuda_codegen.cpp:559]: Added a note to clarify.

File: test/cpp/tensorexpr/test_cuda.cpp (hunk 9 above)
Comment (zheng-xq) [reply test/cpp/tensorexpr/test_cuda.cpp:556]: Added a TODO to add the check in a follow-up.

**Pull request closed without merging**